db/compaction_job.cc

```diff
@@ -1262,18 +1268,18 @@ Status CompactionJob::OpenCompactionOutputFile(
   // Make the output file
   unique_ptr<WritableFile> writable_file;
   EnvOptions opt_env_opts =
-      env_->OptimizeForCompactionTableWrite(env_options_, db_options_);
+      env_->OptimizeForCompactionTableWrite(env_options_, mutable_options_, immutable_options_);
```

Comment:
This is outside db_impl->mutex_, (https://github.com/facebook/rocksdb/blob/master/db/db_impl_compaction_flush.cc#L564-L567) so access mutable_options_ is not safe. I suggest you add a `env_options_for_compaction_` to db_impl, update it in `SetDBOptions()` and pass it to FlushJob and CompactionJob by value, and make sure access of `env_options_for_compaction_` are all in db mutex.